File: framework/csharp/soda/unittest/validate/FeatureFactory.cs
namespace Soda.Unittest.Validate;

public static class FeatureFactory
{
    static FeatureFactory()
    {
        // double
        registerFactory(() => new ObjectFeature<double>(
            (double d) => (long) d.GetHashCode(),
            (double a, double b) => Math.Abs(a - b) < 1e-6
        ));

        // IList<double>
        registerFactory(() => ListFeatureFactory.Ordered<double>(create<double>()));

        // double[]
        registerFactory(() => {
            var proxy = create<IList<double>>();
            return new ObjectFeature<double[]>(
                (double[] arr) => proxy.Hash(arr),
                (double[] x, double[] y) => proxy.IsEqual(x, y)
            );
        });

        // IList<IList<double>>
        registerFactory(() => ListFeatureFactory.Ordered<IList<double>>(create<IList<double>>()));

        // double[][]
        registerFactory(() => {
            var proxy = create<IList<IList<double>>>();
            return new ObjectFeature<double[][]>(
                (double[][] arr2d) => proxy.Hash(arr2d),
                (double[][] x, double[][] y) => proxy.IsEqual(x, y)
            );
        });
    }

    private static readonly IDictionary<Type, Func<ObjectFeature>> factoryMap 
        = new Dictionary<Type, Func<ObjectFeature>>();

    private static void registerFactory<T>(Func<ObjectFeature<T>> factory)
    {
        factoryMap[typeof(T)] = factory;
    }

    public static ObjectFeature<T> create<T>()
    {
        var type = typeof(T);
        if (factoryMap.ContainsKey(type)) {
            return (ObjectFeature<T>) factoryMap[type]();
        }
        return new ObjectFeature<T>(
            (T obj) => (long) obj.GetHashCode(),
            (T a, T b) => object.Equals(a, b)
        );
    }
}

File: framework/kotlin/src/main/kotlin/soda/kotlin/unittest/validate/FeatureFactory.kt
package soda.kotlin.unittest.validate

import kotlin.reflect.KType
import kotlin.reflect.typeOf

object FeatureFactory {

    private val factoryMap = mutableMapOf<String, () -> OFHandle>()

    fun <T> create(type: KType): ObjectFeature<T> {
        val typeName = type.toString()
        return factoryMap[typeName]?.invoke()?.get() ?: CommonFeatureFactory.forDefault()
    }

    inline fun <reified T> create(): ObjectFeature<T> {
        return create(typeOf<T>())
    }

    private inline fun <reified T> registerFactory(noinline factory: () -> ObjectFeature<T>) {
        factoryMap[typeOf<T>().toString()] = { OFHandle(factory()) }
    }

    init {
        registerFactory { CommonFeatureFactory.forDouble() }
        registerFactory { CommonFeatureFactory.forList(create<Double>()) }
        registerFactory { CommonFeatureFactory.forList(create<List<Double>>()) }
        registerFactory { CommonFeatureFactory.forDoubleArray() }
        registerFactory { CommonFeatureFactory.forDoubleArray2d() }
    }
}
